lms/services/digest.py

```diff
@@ -1,11 +1,303 @@
+from dataclasses import dataclass
+from typing import Iterable, Optional, Tuple
+
+from sqlalchemy import func, select
+
+from lms.models import (
+    Assignment,
+    AssignmentGrouping,
+    AssignmentMembership,
+    Course,
+    Grouping,
+    LTIRole,
+    User,
+)
+from lms.services.h_api import HAPI
+from lms.services.mailchimp import EmailRecipient, EmailSender, MailchimpService
+
+
 class DigestService:
     """A service for generating "digests" (activity reports)."""
 
+    def __init__(self, db, h_api, mailchimp_service, sender):
+        self._db = db
+        self._h_api = h_api
+        self._mailchimp_service = mailchimp_service
+        self._sender = sender
```

Comment:
This is pretty much the same as before. The only difference is that previously the factory passed in a `DigestHelper` instance to `DigestService.__init__()`. The factory no longer does that. Instead `send_instructor_email_digests()` constructs a `DigestContext` instance each time it's called, as you'll see below.